### PR TITLE
Alternative Agent API Proposal

### PR DESCRIPTION
**Pull request by MLLeKander**
Comments, criticisms, concerns, or anything beginning with the letter C are welcome.

### CODE REVIEW

File: src/main/java/BallFollower.java [hunk 1]
```diff
@@ -1,19 +1,22 @@
-public class BallFollower implements PlayerInputProvider {
-   public PlayerInput in = new PlayerInput();
-   public GameState game;
-   public Side side;
+public class BallFollower extends AbstractPlayerInputProvider {
```
Comment (Kiskae) [inline src/main/java/BallFollower.java]: Forgot to rename superclass

**Comment by Kiskae** [comment]
Looking at the PR my personal concerns come down to one issue
- Semantics about the definition of 'your player', keeping track of LPlayer and RPlayer and the constant flipping seems like the most unnatural way to talk about the players from the perspective of a controller. I personally prefer the semantic split into 'self()' and 'opponent()' regardless of their side.

Besides that I'm also wondering what your solution to the game loop control requirement for the training phase is. My approach with the GameMonitor (which probably should be named GameSupervisor) was to allow a pluggable controller which could potentially affect the configuration and flow of the game loop.

**Comment by MLLeKander** [comment]
Thanks for the heads up about the refactoring... that's what I get for working so late. :P

I'm not sure what you mean by constant flipping. The flipping should only happen in the `AbstractInputProvider`, before the agent ever sees it.

I'm working on a proposal for the game loop control at the moment that's mostly based on yours. I was going to propose splitting out the construction code into a `GameFactory` class, and leave the monitoring in `GameMonitor`.

**Comment by Kiskae** [comment]
@OMGTallMonster I'm referring to the presence of both `AbstractInputProvider` and `PlayerInputProvider` as well as the public methods `copy()` and `flipX()` (which also copies) on the PhysicsCircle, there should be one consistent way of interacting with the API in my opinion

**Pull request merged**